llama-cpp-2/src/token/data_array.rs

```diff
@@ -47,4 +60,89 @@ impl LlamaTokenDataArray {
     {
         Self::new(data.into_iter().collect(), sorted)
     }
+
+    /// Returns the current selected token, if one exists.
+    #[must_use]
+    pub fn selected_token(&self) -> Option<LlamaToken> {
+        self.data.get(self.selected?).map(LlamaTokenData::id)
+    }
+}
+
+impl LlamaTokenDataArray {
+    /// Modify the underlying data as a `llama_token_data_array`. and reconstruct the `LlamaTokenDataArray`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if some of the safety conditions are not met. (we cannot check all of them at
+    /// runtime so breaking them is UB)
+    ///
+    /// SAFETY:
+    /// The returned array formed by the data pointer and the length must entirely consist of
+    /// initialized token data and the length must be less than the capacity of this array's data
+    /// buffer.
+    /// if the data is not sorted, sorted must be false.
+    pub(crate) unsafe fn modify_as_c_llama_token_data_array<T>(
+        &mut self,
+        modify: impl FnOnce(&mut llama_cpp_sys_2::llama_token_data_array) -> T,
+    ) -> T {
+        let size = self.data.len();
+        let data = self
+            .data
+            .as_mut_ptr()
+            .cast::<llama_cpp_sys_2::llama_token_data>();
+
+        let mut c_llama_token_data_array = llama_cpp_sys_2::llama_token_data_array {
+            data,
+            size,
+            selected: self.selected.and_then(|s| s.try_into().ok()).unwrap_or(-1),
+            sorted: self.sorted,
+        };
+
+        let result = modify(&mut c_llama_token_data_array);
+
+        assert!(
+            c_llama_token_data_array.size <= self.data.capacity(),
+            "Size of the returned array exceeds the data buffer's capacity!"
+        );
+        if !ptr::eq(c_llama_token_data_array.data, data) {
+            ptr::copy(
+                c_llama_token_data_array.data,
+                data,
+                c_llama_token_data_array.size,
+            );
+        }
+        self.data.set_len(c_llama_token_data_array.size);
+
+        self.sorted = c_llama_token_data_array.sorted;
+        self.selected = c_llama_token_data_array
+            .selected
+            .try_into()
+            .ok()
+            .filter(|&s| s < self.data.len());
```

Comment:
when does this filter apply?

Reply:
This is just to ensure that we don't store a value in `selected` that is out of bounds for the array. This will probably only occur if the user applies a selection operation (greedy, dist, mirostat) before applying a filter, which seems unlikely. In the meantime, it make the library more robust in theorey, so I think it is worth keeping.